flang/lib/Optimizer/Transforms/RaiseToAffine.cpp

```diff
@@ -9,40 +9,116 @@
 #include "PassDetail.h"
 #include "flang/Optimizer/Dialect/FIRDialect.h"
 #include "flang/Optimizer/Dialect/FIROps.h"
+#include "flang/Optimizer/Transforms/AffinePromotion.h"
 #include "flang/Optimizer/Transforms/Passes.h"
+#include "mlir/IR/Attributes.h"
+#include "mlir/IR/Visitors.h"
 #include "mlir/Dialect/Affine/IR/AffineOps.h"
 #include "mlir/Dialect/StandardOps/IR/Ops.h"
 #include "mlir/Pass/Pass.h"
 #include "mlir/Transforms/DialectConversion.h"
+#include "llvm/ADT/DenseMap.h"
 #include "llvm/Support/CommandLine.h"
+#define DEBUG_TYPE "flang-affine-promotion"
+#include <iostream>
 
 /// disable FIR to affine dialect conversion
 static llvm::cl::opt<bool>
     disableAffinePromo("disable-affine-promotion",
                        llvm::cl::desc("disable FIR to Affine pass"),
                        llvm::cl::init(false));
 
-using namespace fir;
+namespace fir {
+AffineFunctionAnalysis::AffineFunctionAnalysis(FuncOp funcOp)
+    : functionOp(funcOp) {
+    for (LoopOp op : funcOp.getOps<fir::LoopOp>()) {
+      loopAnalysisMap.try_emplace(op, op);
+    }
+    functionOp.getContext();
+}
+
+AffineLoopAnalysis AffineFunctionAnalysis::getChildLoopAnalysis(fir::LoopOp op) const {
+  auto it = loopAnalysisMap.find_as(op);
+  if (it == loopAnalysisMap.end()) {
+    op.emitError("error in fetching loop analysis during affine promotion\n");
+  } else {
+    return it->getSecond();
+  }
+}
 
-namespace {
+AffineLoopAnalysis::AffineLoopAnalysis(fir::LoopOp op) : legality(analyzeLoop(op)) {}
 
-template <typename FROM>
-class OpRewrite : public mlir::RewritePattern {
-public:
-  explicit OpRewrite(mlir::MLIRContext *ctx)
-      : RewritePattern(FROM::getOperationName(), 1, ctx) {}
+AffineLoopAnalysis::AffineLoopAnalysis(const AffineLoopAnalysis &ala) :
+    step(ala.step), legality(ala.legality) {
 };
 
+bool AffineLoopAnalysis::canPromoteToAffine() {
+  return legality;
+}
+
+bool AffineLoopAnalysis::analyzeLoop(fir::LoopOp op) {
+  return analyzeStep(op.step()) && analyzeMemoryAccess(op);
+}
+
+/// checks for loop step to be either constant or variable which was defined
+///  as constant
+bool AffineLoopAnalysis::analyzeStep(const Value stepValue) {
+  ConstantOp stepDefinition = stepValue.getDefiningOp<ConstantOp>();
+  if (stepDefinition) {
+    Attribute stepAttr = stepDefinition.getValue();
+    if (stepAttr.getKind() == StandardAttributes::Integer) {
+      step = stepAttr.cast<IntegerAttr>().getInt();
+      return true;
+    } else {
+      LLVM_DEBUG(llvm::dbgs()<<"AffineLoopAnalysis: cannot promote loop, step not integer\n";
+                 stepAttr.print(llvm::dbgs()););
+      return false;
+    }
+  } else {
+      LLVM_DEBUG(llvm::dbgs()<<"AffineLoopAnalysis: cannot promote loop, step not constant\n";
+                 if (stepValue.getDefiningOp()) stepValue.getDefiningOp()->print(llvm::dbgs()));
+      return false;
+  }
+}
+
+bool AffineLoopAnalysis::analyzeMemoryAccess(fir::LoopOp loop) {
+  // TODO
```

Comment:
Feel free to use `llvm_unreachable` to crash the compiler when it hits a TODO. Inert comments can give the false impression that something is working or be hard to track down later.